packages/nodejs/src/transmitter.ts

```diff
@@ -0,0 +1,78 @@
+import fs from "fs"
+import path from "path"
+import https from "https"
+import http from "http"
+
+import { Configuration } from "./config"
+import { URL, URLSearchParams } from "url"
+import { HashMap } from "@appsignal/types"
+
+export class Transmitter {
+  #config: Configuration
+  #url: string
+  #data: string
+
+  constructor(url: string, data = "", active = true) {
+    this.#config = new Configuration({ active })
+    this.#url = url
+    this.#data = data
+  }
+
+  public async transmit() {
+    return new Promise<HashMap<any>>((resolve, reject) => {
+      const config_data = this.#config.data
+      const params = new URLSearchParams({
+        api_key: config_data["pushApiKey"] || "",
+        name: config_data["name"] || "",
+        environment: config_data["environment"] || "",
+        hostname: config_data["hostname"] || ""
+      })
+
+      const url = new URL(`${this.#url}?${params.toString()}`)
+      const opts = this.requestOptions(url)
+      const requestModule = url.protocol == "http:" ? http : https
+
+      const request = requestModule.request(opts, (response: any) => {
+        const responseStatus = response.statusCode
+        response.setEncoding("utf8")
+        let responseBody = {}
+
+        response
+          .on("data", (responseData: any) => {
+            try {
+              responseBody = JSON.parse(responseData.toString())
+            } catch (e) {
+              console.error(`Error parsing response data ${responseData}`, e)
+            }
+          })
+          .on("end", () => {
+            if (responseStatus == 200) {
+              resolve({ status: responseStatus, body: responseBody })
+            } else {
+              reject({ status: responseStatus, body: responseBody })
+            }
+          })
+      })
+
+      request.write(this.#data)
+      request.end()
+    })
+  }
+
+  private requestOptions(requestUrl: URL): HashMap<any> {
+    const configData = this.#config.data
+
+    return {
+      method: "POST",
+      protocol: requestUrl.protocol,
+      host: requestUrl.hostname,
+      port: requestUrl.port,
+      path: requestUrl.toString(),
+      headers: {
+        "Content-Type": "application/json",
+        "Content-Length": this.#data.length
+      },
+      cert: fs.readFileSync(configData["caFilePath"] || "", "utf-8")
```

Comment:
Additionally, I'm not entirely sure about this, but going by [the docs for the `tls.createSecureContext` function](https://nodejs.org/api/tls.html#tlscreatesecurecontextoptions), I _think_ the contents of the file, which default to a dump of the Mozilla CAs, should be passed as the `ca` option, not `cert`:

> `ca`: Optionally override the trusted CA certificates. Default is to trust the well-known CAs curated by Mozilla. Mozilla's CAs are completely replaced when CAs are explicitly specified using this option. [...]

> `cert`: Cert chains in PEM format. One cert chain should be provided per private key. Each cert chain should consist of the PEM formatted certificate for a provided private `key`, followed by the PEM formatted intermediate certificates (if any), in order, and not including the root CA (the root CA must be pre-known to the peer, see `ca`). [...]

I believe that, since we're not providing any private keys as `key`, the contents of the CA file that we're passing to `cert` are being ignored. I also believe that those contents would not be valid to be passed to `cert` anyway, as they contain root CAs and do not contain intermediate certificates. See `head packages/nodejs/cert/cacert.pem`:

```
## Bundle of CA Root Certificates
##
## Certificate data from Mozilla as of: Wed Jan  1 04:12:10 2020 GMT
##
## This is a bundle of X.509 certificates of public Certificate Authorities
## (CA). These were automatically extracted from Mozilla's root certificates
## file (certdata.txt).
```

Again, I'm not very confident about any of this. TLS is hard.

Reply:
> Won't this always fail if configData["caFilePath"] is undefined? It seems like fs.readFileSync("", "utf-8") will always throw an ENOENT error.

`configData["caFilePath"]` should never be undefined as it has a default. It'll only be undefined if the users decide it. If that happens, making the script fail is OK to me.

Wow, we were already using `cert` before these changes, now I know why all my tests passed with this. Gonna try using the `ca` option.

Reply:
A good way to test would be to replace `cacert.pem` with an empty file -- if the CAs are being replaced correctly with those on the file, the request should fail.

Reply:
I've tested it and everything passes 🤔

Reply:
Empty files, empty strings, `undefined`, and `null` values are bypassed using the default certificates.

Now we use `ca` instead of `cert`, allowing it to fail if the file has the wrong contents.

Reply:
> configData["caFilePath"] should never be undefined

It can be set by the user (which is why we listen to it here), and also set to an empty value. Which means we should check if the config option is set, the file exists and is a readable file, as documented here: https://github.com/appsignal/integration-guide/pull/104

If I now do this I get an error:

```
$ export APPSIGNAL_CA_FILE_PATH="/foo"
$ bin/diagnose
[...]
  Send diagnostics report to AppSignal? (Y/n): y
  Transmitting diagnostics report

  Error: Something went wrong while submitting the report to AppSignal.
  Response code: undefined
  Response body:
undefined
```

With an empty value:

```
$ export APPSIGNAL_CA_FILE_PATH=""
$ bin/diagnose
[...]
  Send diagnostics report to AppSignal? (Y/n): y
  Transmitting diagnostics report

  Your support token: token
  View this report:   https://appsignal.com/diagnose/token
```

In the Ruby gem, in both scenarios, the config option value is ignored because it's invalid.